### PR TITLE
Animal Phosphorus/Nutrient Refresh

### LINKED ISSUE #1929: [Animal][Nutrient] Nutrient Refactor Implementation [3/3]
Follow the Design in this Google Doc: [Animal Phos Update](https://docs.google.com/document/d/1PAro5--SNBkYvkScfHwWILVAA-jRUU5HRlhQNo8Y0Xw/edit) to implement the refactor of the Nureient submodule that handles the daily phosphorus requirements evaluation and phosphorus updates.

### PR DESCRIPTION
**Pull request by ew3361zh**
Following the structure outlined in the Animal refresh, this PR establishes the AnimalPhosphorusStatus calculator class and the AnimalPhosphorus class to perform the daily phosphorus update for each animal.

### Context

Issue(s) closed by this pull request: closes #1929

- [x] The [changelog](https://docs.google.com/spreadsheets/d/1U-3igxdstHTFb6k5dVcE5-x9IL0r6q18eUiwwytKZyA/edit#gid=0) has been updated.
      
How big are the changes in the PR? Would you nominate it for a major version upgrade?
- [ ] Yes
- [x] No

### What

Initial setup of the AnimalPhosphorus and AnimalPhosphorusStatus classes to match structure designed in #1796.
The AnimalPhosphorusStatus class pulls in the necessary animal attributes and passes them along to the AnimalPhosphorus class where the phosphorus requirements are calculated for each animal based on animal type.

This logic was copied from the current setup but variable names were updated for clarity and readability.
Additionally, the phosphorus requirement calculation and daily update functions from different animal types were combined here within the AnimalPhosphorus class.

Here is a list of the phosphorus-related animal attribute names have been updated for clarity:
`p_excess` -> `phosphorus_excess_in_diet`
`p_intake` -> `phosphorus_intake`
`p_req` -> `phosphorus_requirement`
`dP_reserves` -> `phosphorus_reserves`
`p_animal` -> `total_phosphorus_in_animal`
`p_growth` -> `phosphorus_for_growth`
`p_maint_feces` -> `phosphorus_endogenous_loss`
`p_conc_ration` -> `ration_phosphorus_concentration`
`p_gest` -> `phosphorus_for_gestation`
`p_gest_for_calf` -> `phosphorus_for_gestation_required_for_calf`

### Why

This step is necessary for implementation of the daily phosphorus update within the new Animal structure.

### How

Outside the 2 new classes, the established structure for calculating animal phosphorus requirements and running the daily update was migrated to these new classes.
This is not implemented here - the structure is provided as a framework for once the new animal structure is established.

### Test plan

No testing was added or updated.

### CODE REVIEW

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        self.daily_growth: float = 0.0
+        self.days_in_preg: int = 0
+        self.milking: bool = False
+        self.estimated_daily_milk_produced: float = 0.0
+        self.phosphorus_excess: float = 0.0  # (p_excess)
+        self.phosphorus_intake: float = 0.0  # (p_intake)
+        self.phosphorus_requirement: float = 0.0  # (p_req)
+        self.phosphorus_reserves: float = 0.0  # (dP_reserves)
+        self.total_phosphorus: float = 0.0  # (p_animal)
+        self.phosphorus_retained_for_growth: float = 0.0  # (p_growth)
```
Comment (Rebecavb) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py]: This looks good, just as a suggestion if you want to be consistent with the other names, we could use "phosphorus_for_growth"

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        self.endogenous_loss_phosphorus: float = 0.0  # (p_maint_feces)
```
Comment (Rebecavb) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py]: Same with this one, maybe adding the word "phosphorus" at the beginning will make it more consistent. "phosphorus_endogenous_loss"

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        self.ration_phosphorus_concentration: float = 0.0  # (p_conc_ration)
+        self.gestational_phosphorus: float = 0.0  # (p_gest)
```
Comment (Rebecavb) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py]: Same as the other ones, these names look clear to me but just as a suggestion we could do "phosphorus_for_gestation"

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        self.gestational_phosphorus_for_calf: float = 0.0  # (p_gest_for_calf)
```
Comment (Rebecavb) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py]: Suggestion "phosphorus_for_gestation_required_for_calf"

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    def __init__(self) -> None:
+        """Starting list for attribute data needed from Animal object:"""
+        self.animal_type: AnimalType = None
+        self.body_weight: float = 0.0
+        self.mature_body_weight: float = 0.0
+        self.daily_growth: float = 0.0
+        self.days_in_preg: int = 0
+        self.milking: bool = False
+        self.estimated_daily_milk_produced: float = 0.0
+        self.phosphorus_excess: float = 0.0  # (p_excess)
```
Comment (Rebecavb) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py]: "phosphorus_excess_in_diet"

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        self.phosphorus_intake: float = 0.0  # (p_intake)
+        self.phosphorus_requirement: float = 0.0  # (p_req)
+        self.phosphorus_reserves: float = 0.0  # (dP_reserves)
```
Comment (Rebecavb) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py]: Based on the documentation dP_reserves means: a change in body P reserves on the previous day (g)
I think "phosphorus_reserves" works good

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        self.total_phosphorus: float = 0.0  # (p_animal)
```
Comment (Rebecavb) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py]: "total_phosphorus_in_animal"

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py (hunk 2 above)
Comment (ew3361zh) [reply RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py]: Makes sense, thank you!

**Review by Rebecavb — APPROVED** [review]

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+from RUFAS.routines.animal.animal_types import AnimalType
+
+CALVES_AND_HEIFERS = [AnimalType.CALF, AnimalType.HEIFER_I, AnimalType.HEIFER_II, AnimalType.HEIFER_III]
+HEIFERS_AND_COWS = [AnimalType.DRY_COW, AnimalType.LAC_COW, AnimalType.HEIFER_II, AnimalType.HEIFER_II]
+
+
+class AnimalPhosphorus:
+
+    @staticmethod
+    def perform_daily_phosphorus_update(animal_status: AnimalPhosphorusStatus, animal) -> None:
```
Comment (allisterakun) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py]:
```suggestion
    def perform_daily_phosphorus_update(phosphorus_status: AnimalPhosphorusStatus, general_properties: GeneralPropertiesl) -> AnimalPhosphorusStatus:
```

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+        """Manages animal's daily phosphorus update."""
+        animal_status.get_phosphorus_data(animal)
```
Comment (allisterakun) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py]:
```suggestion
```

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+        animal_status.total_phosphorus_in_animal = (
+            animal_status.total_phosphorus_in_animal
+            + animal_status.phosphorus_for_gestation
+            + animal_status.phosphorus_for_growth
+            + (animal_status.phosphorus_reserves - previous_phosphorus_reserves)
+        )
+
+    @staticmethod
+    def _calculate_phosphorus_requirements(animal_status: AnimalPhosphorusStatus, dry_matter_intake: float) -> None:
```
Comment (allisterakun) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py]:
```suggestion
    def _calculate_phosphorus_requirements(general_properties: GeneralProperties, phosphorus_status: AnimalPhosphorusStatus, dry_matter_intake: float) -> AnimalPhosphorusStatus:
```

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+        ----------
+        RuFaS Phosphorus Animal Module documentation sections A.1A-D.E.1, A.1EF.E.1.
+        """
+        if AnimalPhosphorus._is_calf_or_heifer(animal_status.animal_type):
+            animal_status.phosphorus_endogenous_loss = 0.0008 * dry_matter_intake * GeneralConstants.KG_TO_GRAMS
+        else:
+            animal_status.phosphorus_endogenous_loss = 0.001 * dry_matter_intake * GeneralConstants.KG_TO_GRAMS
+
+    @staticmethod
+    def _calculate_phosphorus_for_growth(animal_status: AnimalPhosphorusStatus) -> None:
```
Comment (allisterakun) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py]:
```suggestion
    def _calculate_phosphorus_for_growth(general_properties: GeneralProperties, phosphorus_status: AnimalPhosphorusStatus) -> float:
```

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+            or animal_status.body_weight < animal_status.mature_body_weight
+        ):
+            animal_status.phosphorus_for_growth = (
+                (0.0012 + 0.004635 * (animal_status.mature_body_weight**0.22) * (animal_status.body_weight ** (-0.22)))
+                * animal_status.daily_growth
+                / 0.96
+                * GeneralConstants.KG_TO_GRAMS
+            )
+        else:
+            animal_status.phosphorus_for_growth = 0.0
```
Comment (allisterakun) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py]:
```suggestion
        if (
            AnimalPhosphorus._is_calf_or_heifer(general_properties.animal_type)
            or general_properties.body_weight < general_properties.mature_body_weight
        ):
            return (
                (0.0012 + 0.004635 * (general_properties.mature_body_weight**0.22) * (general_properties.body_weight ** (-0.22)))
                * general_properties.daily_growth
                / 0.96
                * GeneralConstants.KG_TO_GRAMS
            )
        else:
            return 0.0
```

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+        AnimalPhosphorus._calculate_phosphorus_endogenous_loss(animal_status, dry_matter_intake)
+        urine_production_phosphorus = 0.000002 * animal_status.body_weight * GeneralConstants.KG_TO_GRAMS
+        AnimalPhosphorus._calculate_phosphorus_for_growth(animal_status)
+        AnimalPhosphorus._calculate_gestational_phosphorus(animal_status)
+        milk_phosphorus = AnimalPhosphorus._calculate_milk_phosphorus(animal_status)
+        absorbed_phosphorus = AnimalPhosphorus._calculate_absorbed_phosphorus(
+            animal_status, milk_phosphorus, urine_production_phosphorus
+        )
+        AnimalPhosphorus._calculate_animal_phosphorus_requirement(animal_status, absorbed_phosphorus)
+
```
Comment (allisterakun) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py]:
```suggestion
        phosphorus_status.phosphorus_endogenous_loss  = AnimalPhosphorus._calculate_phosphorus_endogenous_loss(general_properties, phosphorus_status, dry_matter_intake)
        urine_production_phosphorus = 0.000002 * phosphorus_status.body_weight * GeneralConstants.KG_TO_GRAMS
        phosphorus_status.phosphorus_for_growth  = AnimalPhosphorus._calculate_phosphorus_for_growth(general_properties, phosphorus_status)
        phosphorus_status.phosphorus_for_gestation = AnimalPhosphorus._calculate_gestational_phosphorus(general_properties, phosphorus_status)
        phosphorus_status.phosphorus_for_gestation_required_for_calf += phosphorus_status.phosphorus_for_gestation
        milk_phosphorus = AnimalPhosphorus._calculate_milk_phosphorus(general_properties, phosphorus_status)
        absorbed_phosphorus = AnimalPhosphorus._calculate_absorbed_phosphorus(
            general_properties, phosphorus_status, milk_phosphorus, urine_production_phosphorus
        )
        phosphorus_status.phosphorus_requirement  = AnimalPhosphorus._calculate_animal_phosphorus_requirement(general_properties, phosphorus_status, absorbed_phosphorus)
        
        return phosphorus_status

```

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+        """Calculates phosphorus required for endogenous loss based on animal type.
+
+        References
+        ----------
+        RuFaS Phosphorus Animal Module documentation sections A.1A-D.E.1, A.1EF.E.1.
+        """
+        if AnimalPhosphorus._is_calf_or_heifer(animal_status.animal_type):
+            animal_status.phosphorus_endogenous_loss = 0.0008 * dry_matter_intake * GeneralConstants.KG_TO_GRAMS
+        else:
+            animal_status.phosphorus_endogenous_loss = 0.001 * dry_matter_intake * GeneralConstants.KG_TO_GRAMS
```
Comment (allisterakun) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py]:
```suggestion
    @staticmethod
    def _calculate_phosphorus_endogenous_loss(general_properties: GeneralProperties, phosphorus_status: AnimalPhosphorusStatus, dry_matter_intake: float) -> float:
        """Calculates phosphorus required for endogenous loss based on animal type.

        References
        ----------
        RuFaS Phosphorus Animal Module documentation sections A.1A-D.E.1, A.1EF.E.1.
        """
        if AnimalPhosphorus._is_calf_or_heifer(general_properties.animal_type):
            return 0.0008 * dry_matter_intake * GeneralConstants.KG_TO_GRAMS
        else:
            return 0.001 * dry_matter_intake * GeneralConstants.KG_TO_GRAMS
```

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+        """
+        if animal_status.days_in_preg >= 190:
+            exp_1 = (0.05527 - 0.000075 * animal_status.days_in_preg) * animal_status.days_in_preg
+            exp_2 = (0.05527 - 0.000075 * (animal_status.days_in_preg - 1)) * (animal_status.days_in_preg - 1)
+            animal_status.phosphorus_for_gestation = (
+                0.00002743 * math.exp(exp_1) - 0.00002743 * math.exp(exp_2)
+            ) * GeneralConstants.KG_TO_GRAMS
+            animal_status.phosphorus_for_gestation_required_for_calf += animal_status.phosphorus_for_gestation
+        else:
+            animal_status.phosphorus_for_gestation = 0.0
```
Comment (allisterakun) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py]:
```suggestion
    @staticmethod
    def _calculate_gestational_phosphorus(general_properties: GeneralProperties, phosphorus_status: AnimalPhosphorusStatus) -> None:
        """Calculates an animal's gestational phosphorus.

        References
        ----------
        RuFaS Phosphorus Animal Module documentation section A.1C-F.E.4.
        """
        if general_properties.days_in_preg >= 190:
            exp_1 = (0.05527 - 0.000075 * general_properties.days_in_preg) * general_properties.days_in_preg
            exp_2 = (0.05527 - 0.000075 * (general_properties.days_in_preg - 1)) * (general_properties.days_in_preg - 1)
            return (
                0.00002743 * math.exp(exp_1) - 0.00002743 * math.exp(exp_2)
            ) * GeneralConstants.KG_TO_GRAMS
        else:
            return 0.0
```

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+        """Calculates an animal's milk phosphorus.
+
+        References
+        ----------
+        RuFaS Phosphorus Animal Module documentation section A.1E.E.5.
+        """
+        if animal_status.milking:
+            return 0.0009 * animal_status.estimated_daily_milk_produced * GeneralConstants.KG_TO_GRAMS
+        else:
+            return 0.0
```
Comment (allisterakun) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py]:
```suggestion
    @staticmethod
    def _calculate_milk_phosphorus(general_properties: GeneralProperties, phosphorus_status: AnimalPhosphorusStatus) -> float:
        """Calculates an animal's milk phosphorus.

        References
        ----------
        RuFaS Phosphorus Animal Module documentation section A.1E.E.5.
        """
        if general_properties.milking:
            return 0.0009 * general_properties.estimated_daily_milk_produced * GeneralConstants.KG_TO_GRAMS
        else:
            return 0.0
```

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+                + animal_status.phosphorus_endogenous_loss
+                + animal_status.phosphorus_for_growth
+                + animal_status.phosphorus_for_gestation
+            )
+        else:
+            return (
+                urine_production_phosphorus
+                + animal_status.phosphorus_endogenous_loss
+                + animal_status.phosphorus_for_growth
+            )
```
Comment (allisterakun) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py]:
```suggestion
    @staticmethod
    def _calculate_absorbed_phosphorus(
        general_properties: GeneralProperties, phosphorus_status: AnimalPhosphorusStatus, milk_phosphorus: float, urine_production_phosphorus: float
    ) -> float:
        """Calculates absorbed phosphorus based on animal type.

        References
        ----------
        RuFaS Phosphorus Animal Module documentation sections A.1EF.E.6, A.1A-F.E.6.
        """
        if general_properties.animal_type in [AnimalType.DRY_COW, AnimalType.LAC_COW]:
            return (
                urine_production_phosphorus
                + phosphorus_status.phosphorus_endogenous_loss
                + phosphorus_status.phosphorus_for_growth
                + phosphorus_status.phosphorus_for_gestation
                + milk_phosphorus
            )
        elif general_properties.animal_type in [AnimalType.HEIFER_II, AnimalType.HEIFER_III]:
            return (
                urine_production_phosphorus
                + phosphorus_status.phosphorus_endogenous_loss
                + phosphorus_status.phosphorus_for_growth
                + phosphorus_status.phosphorus_for_gestation
            )
        else:
            return (
                urine_production_phosphorus
                + phosphorus_status.phosphorus_endogenous_loss
                + phosphorus_status.phosphorus_for_growth
            )
```

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+            if animal_status.milking:
+                animal_status.phosphorus_requirement = absorbed_phosphorus / (
+                    1.86696
+                    - 5.01238 * animal_status.ration_phosphorus_concentration
+                    + 5.12286 * animal_status.ration_phosphorus_concentration**2
+                )
+        elif animal_status.animal_type == AnimalType.CALF:
+            animal_status.phosphorus_requirement = absorbed_phosphorus / 0.90
+        else:
+            animal_status.phosphorus_requirement = absorbed_phosphorus / 0.664
```
Comment (allisterakun) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py]:
```suggestion
    @staticmethod
    def _calculate_animal_phosphorus_requirement(
        general_properties: GeneralProperties, phosphorus_status: AnimalPhosphorusStatus, absorbed_phosphorus: float
    ) -> float:
        """Calculates an animal's phosphorus requirement by animal type.

        References
        ----------
        RuFaS Phosphorus Animal Module documentation sections A.1A.E.7, A.1B-D.E.7, A.1EF.E.7.
        """
        if general_properties.animal_type in [AnimalType.DRY_COW, AnimalType.LAC_COW]:
            if general_properties.milking:
                return absorbed_phosphorus / (
                    1.86696
                    - 5.01238 * phosphorus_status.ration_phosphorus_concentration
                    + 5.12286 * phosphorus_status.ration_phosphorus_concentration**2
                )
        elif general_properties.animal_type == AnimalType.CALF:
            return absorbed_phosphorus / 0.90
        else:
           return absorbed_phosphorus / 0.664
```

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+        else:
+            animal_status.phosphorus_reserves = 0.0
+
+        animal_status.total_phosphorus_in_animal = (
+            animal_status.total_phosphorus_in_animal
+            + animal_status.phosphorus_for_gestation
+            + animal_status.phosphorus_for_growth
+            + (animal_status.phosphorus_reserves - previous_phosphorus_reserves)
+        )
+
```
Comment (allisterakun) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py]:
```suggestion
    @staticmethod
    def _calculate_total_animal_phosphorus(phosphorus_status: AnimalPhosphorusStatus) -> AnimalPhosphorusStatus:
        """Calculates the total phosphorus for the animal.

        References
        ----------
        RuFaS Phosphorus Animal Module documentation sections A.1G.A.1, A.1G.A.2, A.1G.A.3.

        Notes
        -----
        - Change in body P reserves (g), must be <= 0

        """
        phosphorus_status.phosphorus_excess_in_diet = max(
            phosphorus_status.phosphorus_intake - phosphorus_status.phosphorus_requirement, 0
        )
        previous_phosphorus_reserves = phosphorus_status.phosphorus_reserves

        if phosphorus_status.phosphorus_intake < phosphorus_status.phosphorus_requirement:
            phosphorus_status.phosphorus_reserves = (
                phosphorus_status.phosphorus_intake
                - phosphorus_status.phosphorus_requirement
                + phosphorus_status.phosphorus_reserves
            )
        elif (
            phosphorus_status.phosphorus_intake >= phosphorus_status.phosphorus_requirement
            and phosphorus_status.phosphorus_reserves < 0
        ):
            phosphorus_status.phosphorus_reserves = (
                0.7 * phosphorus_status.phosphorus_excess_in_diet + phosphorus_status.phosphorus_reserves
            )
        else:
            phosphorus_status.phosphorus_reserves = 0.0

        phosphorus_status.total_phosphorus_in_animal = (
            phosphorus_status.total_phosphorus_in_animal
            + phosphorus_status.phosphorus_for_gestation
            + phosphorus_status.phosphorus_for_growth
            + (phosphorus_status.phosphorus_reserves - previous_phosphorus_reserves)
        )
        
        return phosphorus_status

```

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+class AnimalPhosphorus:
+
+    @staticmethod
+    def perform_daily_phosphorus_update(animal_status: AnimalPhosphorusStatus, animal) -> None:
+        """Manages animal's daily phosphorus update."""
+        animal_status.get_phosphorus_data(animal)
+        dry_matter_intake = AnimalPhosphorus._get_dry_matter_intake()
+        AnimalPhosphorus._calculate_phosphorus_requirements(animal_status, dry_matter_intake)
+        AnimalPhosphorus._calculate_total_animal_phosphorus(animal_status)
+        animal_status._update_animal(animal)
```
Comment (allisterakun) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus.py]:
```suggestion
        phosphorus_status = AnimalPhosphorus._calculate_phosphorus_requirements(general_properties, phosphorus_status , dry_matter_intake)
        phosphorus_status = AnimalPhosphorus._calculate_total_animal_phosphorus(phosphorus_status )
        
        return phosphorus_status
```

File: RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+    def _update_animal(self, animal) -> None:
+        """Updates the animal's phosphorus-related attributes after the daily routine."""
+        animal.phosphorus_endogenous_loss = self.phosphorus_endogenous_loss
+        animal.phosphorus_for_growth = self.phosphorus_for_growth
+        animal.phosphorus_requirement = self.phosphorus_requirement
+        animal.phosphorus_for_gestation = self.phosphorus_for_gestation
+        animal.phosphorus_for_gestation_required_for_calf = self.phosphorus_for_gestation_required_for_calf
+        animal.phosphorus_excess_in_diet = self.phosphorus_excess_in_diet
+        animal.phosphorus_reserves = self.phosphorus_reserves
+        animal.total_phosphorus_in_animal = self.total_phosphorus_in_animal
```
Comment (allisterakun) [inline RUFAS/routines/animal/animal_nutrients_daily_update/animal_phosphorus_status.py]:
```suggestion
```

**Review by allisterakun — APPROVED** [review]
A lot of the functions are relying on side-effects. Made some suggestions to fix that.

**Pull request merged**